[Core] Reduce number of file watchers used with .NET Core projects  …

mrward opened the pull request:
A file watcher was being created for each SDK style project. When a
solution contains a lot of SDK style projects (e.g. OrchardCore
contains 126 projects) this can then cause Mono's networking stack
to stop working. NuGet restore, check for updates would fail with
errors similar to:

SecureChannelFailure (Unable to write data to the transport connection:
The socket is not connected.)

Since the FileWatcherService has a maximum number file watchers the
project can use the common file watchers instead of creating its own.
This reduces the number of file watchers. For the OrchardCore solution
this reduces the number of file watchers down from 127 to 1.

There are still other NuGet restore problems with the OrchardCore
solution, but this solves one part.

Fixes VSTS #678422 - Unable to restore packages in OrchardCore app

mrward committed:
[Core] Reduce number of file watchers used with .NET Core projects

A file watcher was being created for each SDK style project. When a
solution contains a lot of SDK style projects (e.g. OrchardCore
contains 126 projects) this can then cause Mono's networking stack
to stop working. NuGet restore, check for updates would fail with
errors similar to:

SecureChannelFailure (Unable to write data to the transport connection:
The socket is not connected.)

Since the FileWatcherService has a maximum number file watchers the
project can use the common file watchers instead of creating its own.
This reduces the number of file watchers. For the OrchardCore solution
this reduces the number of file watchers down from 127 to 1.

There are still other NuGet restore problems with the OrchardCore
solution, but this solves one part.

Fixes VSTS #678422 - Unable to restore packages in OrchardCore app

mrward committed:
[Core] Handle directory moved outside project directory outside IDE

SDK style projects that monitor directory changes now support
a directory being moved out of the project's directory, and a
directory being moved into the project's directory, outside the
IDE. Also added a check to ensure directory renames that are
completely outside the project's directory are ignored. Using the
common FileWatcherService means that rename events for all monitored
directories are processed by the project.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -4425,13 +4422,59 @@ void OnFileDeleted (object sender, FileSystemEventArgs e)
+
+		void OnDirectoryMovedIntoProject (FilePath newDirectory)
+		{
+			foreach (string file in Directory.GetFiles (newDirectory, "*", SearchOption.AllDirectories)) {
```
`Directory.EnumerateFiles` here, as `GetFiles` is `EnumerateFiles ().ToArray()`

mrward replied:
Yeah, forgot about EnumerateFiles 👍

Therzok commented on main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/DotNetCoreFileWatcherTests.cs:
```diff
@@ -72,6 +72,8 @@ async Task<DotNetProject> OpenProject (string solutionName = "DotNetCoreFileWatc
 
 			RunMSBuild ($"/t:Restore /p:RestoreDisableParallel=true \"{solution.FileName}\"");
 
+			await FileWatcherService.Add (solution);
```
Where is the filewatcher remove for this? We might have tests hang if we don't remove it.

mrward replied:
Happens when the Solution is disposed. No need for an explicit remove.

Therzok reviewed (approved):
Other than enumeration change, LGTM! 👍

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -4351,47 +4351,44 @@ void InitializeFileWatcher ()
-			watcher.Renamed += OnFileRenamed;
-			watcher.Error += OnFileWatcherError;
-			watcher.EnableRaisingEvents = true;
+			// Use FileWatcherService for file created event since this does not run on the UI thread.
```
It can hang the IDE if there are lot of files being created. Seen this problem when doing an npm install with the project open in the IDE.

slluis replied:
If we need to be able to handle file change events in the same thread that originated the event, let's add that capability to FileService. Having file change events in two classes is confusing.

mrward committed:
[Core] Move file watcher events to FileService

FileService now has an AsyncEvents class with FileCreated, FileRemoved
and FileRenamed events that are fired by the file watcher and do not
occur on the UI thread. The file watcher service events have been
removed.

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/FileService.cs:
```diff
@@ -845,6 +845,11 @@ static void OnFileChanged (FileEventArgs args)
+		/// <summary>
+		/// File watcher events - these are not fired on the UI thread.
+		/// </summary>
+		public static AsyncEvents AsyncEvents = new AsyncEvents ();
```
This should be a property.

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/FileService.cs:
```diff
@@ -966,5 +971,27 @@ public void Thaw ()
+		public event EventHandler<FileEventArgs> FileRemoved;
+		public event EventHandler<FileCopyEventArgs> FileRenamed;
+
+		internal void NotifyFileCreated (FilePath fileName)
```
We don't need Notify* methods in this class. The idea is that the NotifyFileCreated method in FileService fires the FileService.AsyncEvents.FileCreated event from whatever thread is calling NotifyFileCreated, and it fires FileService.FileCreated in the UI thread.


 events defined in AsyncEvents are raised in the thread that notified then.

slluis commented:
@monojenkins backport release-7.7
